# Import libraries

In [1]:
import json         
from datetime import datetime
import pandas as pd
import cProfile

from q1_time import q1_time

---

# Open file

Taking a quick look at the data file, we can see that it consists of a collection of json objects, as shows the image bellow.

![img](../img/data_file.png "screenshot data file")

The amount of data sums up to 117407 lines of json objects.

## Get data

To work with the data we have to extract it from the json file. To do that will run the follwing lines of code.

In [2]:
file_path = "farmers-protest-tweets-2021-2-4.json"

df = pd.read_json(file_path, lines=True)
df.head(10)

url  \
0  https://twitter.com/ArjunSinghPanam/status/136...   
1  https://twitter.com/PrdeepNain/status/13645062...   
2  https://twitter.com/parmarmaninder/status/1364...   
3  https://twitter.com/anmoldhaliwal/status/13645...   
4  https://twitter.com/KotiaPreet/status/13645061...   
5  https://twitter.com/babli_708/status/136450612...   
6  https://twitter.com/Varinde17354019/status/136...   
7  https://twitter.com/BitnamSingh/status/1364505...   
8  https://twitter.com/anmoldhaliwal/status/13645...   
9  https://twitter.com/SatThiara/status/136450589...   

                       date  \
0 2021-02-24 09:23:35+00:00   
1 2021-02-24 09:23:32+00:00   
2 2021-02-24 09:23:22+00:00   
3 2021-02-24 09:23:16+00:00   
4 2021-02-24 09:23:10+00:00   
5 2021-02-24 09:23:05+00:00   
6 2021-02-24 09:22:54+00:00   
7 2021-02-24 09:22:35+00:00   
8 2021-02-24 09:22:34+00:00   
9 2021-02-24 09:22:11+00:00   

                                             content  \
0  The world progresses while the Indian police a...   
1  #FarmersProtest \n#ModiIgnoringFarmersDeaths \...   
2  ਪੈਟਰੋਲ ਦੀਆਂ ਕੀਮਤਾਂ ਨੂੰ ਮੱਦੇਨਜ਼ਰ ਰੱਖਦੇ ਹੋਏ \nਮੇ...   
3  @ReallySwara @rohini_sgh watch full video here...   
4  #KisanEktaMorcha #FarmersProtest #NoFarmersNoF...   
5  Jai jwaan jai kissan #FarmersProtest #ModiIgno...   
6                                    #FarmersProtest   
7  #ModiDontSellFarmers\n#FarmersProtest https://...   
8  @mandeeppunia1 watch full video here https://t...   
9            #FarmersProtest https://t.co/ehd5FBSZGx   

                                     renderedContent                   id  \
0  The world progresses while the Indian police a...  1364506249291784198   
1  #FarmersProtest \n#ModiIgnoringFarmersDeaths \...  1364506237451313155   
2  ਪੈਟਰੋਲ ਦੀਆਂ ਕੀਮਤਾਂ ਨੂੰ ਮੱਦੇਨਜ਼ਰ ਰੱਖਦੇ ਹੋਏ \nਮੇ...  1364506195453767680   
3  @ReallySwara @rohini_sgh watch full video here...  1364506167226032128   
4  #KisanEktaMorcha #FarmersProtest #NoFarmersNoF...  1364506144002088963   
5  Jai jwaan jai kissan #FarmersProtest #ModiIgno...  1364506120497360896   
6                                    #FarmersProtest  1364506076272496640   
7  #ModiDontSellFarmers\n#FarmersProtest twitter....  1364505995859423234   
8  @mandeeppunia1 watch full video here youtu.be/...  1364505991887347714   
9        #FarmersProtest twitter.com/borisjohnson/s…  1364505896576053248   

                                                user  \
0  {'username': 'ArjunSinghPanam', 'displayname':...   
1  {'username': 'PrdeepNain', 'displayname': 'Pra...   
2  {'username': 'parmarmaninder', 'displayname': ...   
3  {'username': 'anmoldhaliwal', 'displayname': '...   
4  {'username': 'KotiaPreet', 'displayname': 'Pre...   
5  {'username': 'babli_708', 'displayname': 'Babl...   
6  {'username': 'Varinde17354019', 'displayname':...   
7  {'username': 'BitnamSingh', 'displayname': 'Bi...   
8  {'username': 'anmoldhaliwal', 'displayname': '...   
9  {'username': 'SatThiara', 'displayname': 'Sat ...   

                                            outlinks  \
0  [https://twitter.com/ravisinghka/status/136415...   
1                                                 []   
2                                                 []   
3                     [https://youtu.be/-bUKumwq-J8]   
4                                                 []   
5  [https://twitter.com/rajeshpunia15/status/1364...   
6                                                 []   
7  [https://twitter.com/jagjitvaheguru/status/136...   
8                     [https://youtu.be/-bUKumwq-J8]   
9  [https://twitter.com/borisjohnson/status/13642...   

                 tcooutlinks  replyCount  retweetCount  ...  quoteCount  \
0  [https://t.co/es3kn0IQAF]           0             0  ...           0   
1                         []           0             0  ...           0   
2                         []           0             0  ...           0   
3  [https://t.co/wBPNdJdB0n]           0             0  ...           0   
4               

In [3]:
df.tail(10)

url  \
117397  https://twitter.com/rupindr79/status/136004022...   
117398  https://twitter.com/bali_mandeep/status/136004...   
117399  https://twitter.com/Satveer22950341/status/136...   
117400  https://twitter.com/PushpSamra/status/13600401...   
117401  https://twitter.com/lovehazran1/status/1360040...   
117402  https://twitter.com/rickyrickstir/status/13600...   
117403  https://twitter.com/PunjabTak/status/136004014...   
117404  https://twitter.com/ish_kayy/status/1360040134...   
117405  https://twitter.com/TV9Bharatvarsh/status/1360...   
117406  https://twitter.com/SikhVibes/status/136004012...   

                            date  \
117397 2021-02-12 01:37:13+00:00   
117398 2021-02-12 01:37:12+00:00   
117399 2021-02-12 01:37:06+00:00   
117400 2021-02-12 01:37:05+00:00   
117401 2021-02-12 01:37:04+00:00   
117402 2021-02-12 01:37:02+00:00   
117403 2021-02-12 01:36:53+00:00   
117404 2021-02-12 01:36:50+00:00   
117405 2021-02-12 01:36:49+00:00   
117406 2021-02-12 01:36:49+00:00   

                                                  content  \
117397  Now Farmers Agitation is no longer confined to...   
117398            Kisan Ekta Zindabaad ✊\n#FarmersProtest   
117399       #FarmersProtest \n\n#MahapanchayatRevolution   
117400  The first Mahapanchayat of Punjab. The revolut...   
117401  #BJPGovtDictatingTwitter #MahapanchayatRevolut...   
117402  #FarmersProtest #KisanAndolan #KisaanMajdoorEk...   
117403  PM मोदी की अपील के बीच संयुक्त किसान मोर्चा का...   
117404  United we stand.\nDivided we fall\n#Mahapancha...   
117405  सिंघु बॉर्डर पर लंबी लड़ाई की तैयारी, किसानों ...   
117406  @Kisanektamorcha We are with you, keep the mor...   

                                          renderedContent  \
117397  Now Farmers Agitation is no longer confined to...   
117398            Kisan Ekta Zindabaad ✊\n#FarmersProtest   
117399       #FarmersProtest \n\n#MahapanchayatRevolution   
117400  The first Mahapanchayat of Punjab. The revolut...   
117401  #BJPGovtDictatingTwitter #MahapanchayatRevolut...   
117402  #FarmersProtest #KisanAndolan #KisaanMajdoorEk...   
117403  PM मोदी की अपील के बीच संयुक्त किसान मोर्चा का...   
117404  United we stand.\nDivided we fall\n#Mahapancha...   
117405  सिंघु बॉर्डर पर लंबी लड़ाई की तैयारी, किसानों ...   
117406  @Kisanektamorcha We are with you, keep the mor...   

                         id  \
117397  1360040229265022979   
117398  1360040222986178563   
117399  1360040199493799936   
117400  1360040195786067969   
117401  1360040192090857475   
117402  1360040182771163138   
117403  1360040146402373637   
117404  1360040134230556678   
117405  1360040127679000577   
117406  1360040127146430470   

                                                     user  \
117397  {'username': 'rupindr79', 'displayname': 'ਰੁ ਪ...   
117398  {'username': 'bali_mandeep', 'displayname': 'M...   
117399  {'username': 'Satveer22950341', 'displayname':...   
117400  {'username': 'PushpSamra', 'displayname': 'Pus...   
117401  {'username': 'lovehazran1', 'displayname': '#F...   
117402  {'username': 'rickyrickstir', 'displayname': '...   
117403  {'username': 'PunjabTak', 'displayname': 'Punj...   
117404  {'username': 'ish_kayy', 'displayname': 'ishy'...   
117405  {'username': 'TV9Bharatvarsh', 'displayname': ...   
117406  {'username': 'SikhVibes', 'displayname': 'Sikh...   

                                                 outlinks  \
117397                                                 []   
117398                                                 []   
117399                                                 []   
117400                                                 []   
117401                                                 []   
117402                                                 []   
117403                     [https://youtu.be/aG3qHGwoYag]   
117404                                                 []   
117405  [https://www.tv9hindi.com/india/farmers-protes...   
117406           

In [4]:
shape = df.shape
print("Rows: ", shape[0])
print("Columns: ", df.columns)
print("Columns count: ", shape[1])

Rows:  117407
Columns:  Index(['url', 'date', 'content', 'renderedContent', 'id', 'user', 'outlinks',
       'tcooutlinks', 'replyCount', 'retweetCount', 'likeCount', 'quoteCount',
       'conversationId', 'lang', 'source', 'sourceUrl', 'sourceLabel', 'media',
       'retweetedTweet', 'quotedTweet', 'mentionedUsers'],
      dtype='object')
Columns count:  21


We have 117407 lines with 21 columns in the dataframe

In [5]:
df.isna().sum()

url                     0
date                    0
content                 0
renderedContent         0
id                      0
user                    0
outlinks                0
tcooutlinks             0
replyCount              0
retweetCount            0
likeCount               0
quoteCount              0
conversationId          0
lang                    0
source                  0
sourceUrl             912
sourceLabel           912
media               89298
retweetedTweet     117407
quotedTweet         75971
mentionedUsers      79373
dtype: int64

We can see that `media`, `retweetedTweet`, `quotedTweet` and `mentionedUsers` have a relevant amount of `None` value in its columns. As we answer the questions, we'll avaluate if it will impact us or just ignore them.

---

In [6]:
df.dtypes

url                             object
date               datetime64[ns, UTC]
content                         object
renderedContent                 object
id                               int64
user                            object
outlinks                        object
tcooutlinks                     object
replyCount                       int64
retweetCount                     int64
likeCount                        int64
quoteCount                       int64
conversationId                   int64
lang                            object
source                          object
sourceUrl                       object
sourceLabel                     object
media                           object
retweetedTweet                 float64
quotedTweet                     object
mentionedUsers                  object
dtype: object

# Question 1

**Las top 10 fechas donde hay más tweets. Mencionar el usuario (username) que más publicaciones tiene por cada uno de esos días. Debe incluir las siguientes funciones**

```python
Returns: 
[(datetime.date(1999, 11, 15), "LATAM321"), (datetime.date(1999, 7, 15), "LATAM_CHI"), ...]
```


Let's take the coulmns we're interested in and work it out from there. For this question we'll isolate `date` and `user` columns in a separete dataframe to not mess up with the original data.

In [7]:
df_q1 = df[["date", "user"]]
df_q1.head(5)

date                                               user
0 2021-02-24 09:23:35+00:00  {'username': 'ArjunSinghPanam', 'displayname':...
1 2021-02-24 09:23:32+00:00  {'username': 'PrdeepNain', 'displayname': 'Pra...
2 2021-02-24 09:23:22+00:00  {'username': 'parmarmaninder', 'displayname': ...
3 2021-02-24 09:23:16+00:00  {'username': 'anmoldhaliwal', 'displayname': '...
4 2021-02-24 09:23:10+00:00  {'username': 'KotiaPreet', 'displayname': 'Pre...

Comparing the inputs to the expected output, it's clear that the values in the dataframe need to be processed first.

Starting with the `date` column. Taking the date on index 0, we have `2021-02-24T09:23:35+00:00`, in general terms: YYYY-MM-DD HH:MM:SS UTC time. So what needs to be done is, convert the date values on the `date` column into a datetime object. 

In [8]:
df_q1["date"] = df_q1['date'].dt.date
df_q1.head(5)

/var/folders/zt/dj3z9scd4m3dpw0vbtj9fbm80000gn/T/ipykernel_20281/2127940491.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_q1["date"] = df_q1['date'].dt.date


date                                               user
0  2021-02-24  {'username': 'ArjunSinghPanam', 'displayname':...
1  2021-02-24  {'username': 'PrdeepNain', 'displayname': 'Pra...
2  2021-02-24  {'username': 'parmarmaninder', 'displayname': ...
3  2021-02-24  {'username': 'anmoldhaliwal', 'displayname': '...
4  2021-02-24  {'username': 'KotiaPreet', 'displayname': 'Pre...

Furthermore, we have to process the `user` column to filter information we are interested in: the value of `username` key.

In [9]:
df_q1["user"] = df_q1.apply(lambda x: x.user["username"], axis=1)
df_q1.head(5)

/var/folders/zt/dj3z9scd4m3dpw0vbtj9fbm80000gn/T/ipykernel_20281/1694884845.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_q1["user"] = df_q1.apply(lambda x: x.user["username"], axis=1)


date             user
0  2021-02-24  ArjunSinghPanam
1  2021-02-24       PrdeepNain
2  2021-02-24   parmarmaninder
3  2021-02-24    anmoldhaliwal
4  2021-02-24       KotiaPreet

Now, lets first verify the total amount of tweets made in all the dates. Then, will sort the results and tke the top 10 of them.

In [10]:
grouped_date = df_q1.groupby("date", as_index=False)
top_dates = grouped_date.count().sort_values(by=["user"], ascending=False).head(10)
top_dates

date   user
0   2021-02-12  12347
1   2021-02-13  11296
5   2021-02-17  11087
4   2021-02-16  10443
2   2021-02-14  10249
6   2021-02-18   9625
3   2021-02-15   9197
8   2021-02-20   8502
11  2021-02-23   8417
7   2021-02-19   8204

With the above results, we know the list of dates that the answer must contain.

With the same idea, we'll group `date` and `username` coulmns together.

In [11]:
grouped_user = df_q1.groupby(["date", "user"], as_index=False).size().sort_values(by="size", ascending=False)
grouped_user.head(10)

date             user  size
35219  2021-02-19         Preetm91   267
33193  2021-02-18  neetuanjle_nitu   195
26577  2021-02-17   RaaJVinderkaur   185
7536   2021-02-13  MaanDee08215437   178
2740   2021-02-12  RanbirS00614606   176
42691  2021-02-21       Surrypuria   161
33396  2021-02-18    rebelpacifist   153
34733  2021-02-19  KaurDosanjh1979   138
48696  2021-02-23       Surrypuria   135
18540  2021-02-15           jot__b   134

We've got to be careful now. It might look like the final answer, but what the output really shows is most tweeted dates made by a single user. Witch is different from what we are looking for.

So, now we have to filter the dates based on the results showed at `top_dates` dataframe, sort the results then drop the duplicated dates.

In [12]:
dates = top_dates.date.to_list()
top_users = df_q1.groupby(["date", "user"], as_index=False).size().set_index("date").loc[dates].sort_values(by="size", ascending=False).reset_index().drop_duplicates(subset=["date"])
top_users

date             user  size
0   2021-02-19         Preetm91   267
1   2021-02-18  neetuanjle_nitu   195
2   2021-02-17   RaaJVinderkaur   185
3   2021-02-13  MaanDee08215437   178
4   2021-02-12  RanbirS00614606   176
7   2021-02-23       Surrypuria   135
8   2021-02-15           jot__b   134
9   2021-02-16           jot__b   133
12  2021-02-14    rebelpacifist   119
23  2021-02-20  MangalJ23056160   108

We can see that the two previous results are slightly different from each another. If you look closely, you'll notice that `2021-02-21` is in the first but not int the seconde answer.

Furthermore, we'll sort the answer by the most to less tweeted ones.

In [13]:
top_dates_dict = top_dates.set_index("date").to_dict()

q1_result_df = top_users.copy()
q1_result_df["tweets"] = [top_dates_dict["user"][d] for d in top_users.date]
q1_result_df = q1_result_df.sort_values(by="tweets", ascending=False).reset_index(drop=True)
q1_result_df

date             user  size  tweets
0  2021-02-12  RanbirS00614606   176   12347
1  2021-02-13  MaanDee08215437   178   11296
2  2021-02-17   RaaJVinderkaur   185   11087
3  2021-02-16           jot__b   133   10443
4  2021-02-14    rebelpacifist   119   10249
5  2021-02-18  neetuanjle_nitu   195    9625
6  2021-02-15           jot__b   134    9197
7  2021-02-20  MangalJ23056160   108    8502
8  2021-02-23       Surrypuria   135    8417
9  2021-02-19         Preetm91   267    8204

Lastly, we can have turn the answer into the right form.

In [14]:
date = q1_result_df["date"]
user = q1_result_df["user"]
q1_result = [(date.iloc[i], user.iloc[i]) for i in range(10)]
q1_result

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'rebelpacifist'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 15), 'jot__b'),
 (datetime.date(2021, 2, 20), 'MangalJ23056160'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 19), 'Preetm91')]

## Q1 Memory Usage

In [15]:
!python -m memory_profiler q1_memory.py "farmers-protest-tweets-2021-2-4.json"

Filename: q1_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     8     76.9 MiB     76.9 MiB           1   @profile
     9                                         def q1_memory(file_path: str) -> List[Tuple[datetime.date, str]]:
    10                                         
    11                                             # Get data from json file
    12   1554.7 MiB   1477.8 MiB           1       df_q1 = pd.read_json(file_path, lines=True)[["date", "user"]]
    13                                         
    14                                             # Transform the values in the columns
    15   1543.1 MiB    -11.6 MiB           1       df_q1["date"] = df_q1['date'].dt.date
    16   1548.1 MiB   -279.9 MiB      234815       df_q1["user"] = df_q1.apply(lambda x: x.user["username"], axis=1)
    17                                         
    18                                             # Most tweeted dates
    19   1263.4 MiB   -284.7 MiB           1

It's clear that the data that, ,in this case, is saved in a dataframe, takes a great part of the memory usage.

An improvement to be made is to search for another method to store data. It has to be light weight and simple to be used as pandas dataframes.

## Q1 Time Consumption

In [16]:
cProfile.run("q1_time(file_path)")

         4755643 function calls (4755009 primitive calls) in 22.540 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(all)
       10    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(append)
        9    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(argsort)
        9    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(atleast_2d)
        6    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(bincount)
       34    0.000    0.000    0.281    0.008 <__array_function__ internals>:177(concatenate)
       92    0.000    0.000    0.001    0.000 <__array_function__ internals>:177(copyto)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(delete)
        5    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(insert)
        5  

We can see that the total time consumed to generate the answer to the question 1 was 22.540 seconds.

We can see that a lot of time was dedicated to read the json file. Another time consumming task is the `apply` method.

As mentioned earlier, dataframes might not be the best way to handle the data as it should. A better option would be to look for a tool suitable for big data. The one should be light weight, easy to use and fast.

---